Implement Masonry Track Sizing

stwrt opened the pull request:
#### 3c5823bb309e94722ae6028f80166daa2d448f4e
<pre>
Implement Masonry Track Sizing
<a href="https://bugs.webkit.org/show_bug.cgi?id=248287">https://bugs.webkit.org/show_bug.cgi?id=248287</a>

Reviewed by Matt Woodrow.

Track sizing will only include blocks if they meet the below criteria in the track sizing algorithm.

- Items placed at the first implicit line in the masonry axis.
- Items that have a specified definite placement in the grid axis
- Items that span all grid axis tracks.

The masonry calling code needed to be heavily refactored to a later step in layout pipeline.
This ended breaking quite a few test cases, which will be addressed in future patches.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/track-sizing/masonry-track-sizing-overflow-left-side-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/track-sizing/masonry-track-sizing-overflow-left-side.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/track-sizing/masonry-track-sizing-overflow-right-side-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/track-sizing/masonry-track-sizing-overflow-right-side.html: Added.
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/track-sizing/masonry-track-sizing-overflow-left-side-expected.txt: Added.
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/track-sizing/masonry-track-sizing-overflow-right-side-expected.txt: Added.
* Source/WebCore/rendering/GridMasonryLayout.h:
(WebCore::GridMasonryLayout::gridGap const):
* Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
(WebCore::GridTrackSizingAlgorithm::accumulateIntrinsicSizesForTrack):
(WebCore::GridTrackSizingAlgorithm::run):
* Source/WebCore/rendering/RenderGrid.cpp:
(WebCore::RenderGrid::layoutBlock):
(WebCore::RenderGrid::placeItemsOnGrid):
(WebCore::RenderGrid::autoPlacementMinorAxisDirection const):

Canonical link: <a href="https://commits.webkit.org/258280@main">https://commits.webkit.org/258280@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/52833274df91392baf1bc1fd572b44308ae86e60

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/101390 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/10548 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/34448 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/110662 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/170929 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/105371 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/11498 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/1401 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/93785 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/108488 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/107172 "Passed tests") | [⏳ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/iOS-16-Simulator-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/91984 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/35269 "Passed tests") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/23391 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/78266 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/4160 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/24903 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/4215 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/1319 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/10310 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/44391 "Passed tests") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5693 "Failed to push commit to Webkit repository") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/5979 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

mattwoodrow commented on LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/track-sizing/masonry-track-sizing-overflow-left-side.html:
```diff
@@ -0,0 +1,56 @@
+    <title>CSS Grid Test: Verify correct behavior of track-sizing</title>
+    <link rel="author" title="Brandon Stewart" href="mailto:brandonstewart@apple.com">
+    <link rel="help" href="https://drafts.csswg.org/css-grid-3/#track-sizing">
+    <link rel="match" href="masonry-track-sizing.html">
```
"masonry-track-sizing-overflow-left-side-ref.html"

You also need to duplicate the -ref file with a -expected suffix. That's because our test harness (unlike the official WPT harness) ignores the "match" header, and just unconditionally looks for a file matching the test file + "-expected".

When upstreaming to the WPT repo, only upstream the -ref file, not the -expected.

If you do that, the -expected.txt files you've added should be removed, and it'll be using the -expected.html versions instead.

stwrt replied:
Fixed.

mattwoodrow commented on LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/track-sizing/masonry-track-sizing-overflow-right-side.html:
```diff
@@ -0,0 +1,58 @@
+}
+
+box3 {
+    width: 1fr;
```
The 'fr' unit is a special thing that can only be used in the template of grid tracks, it's not meaningful to use it to size other boxes.

stwrt replied:
Thanks. Deleted.

mattwoodrow commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1603,6 +1634,12 @@ void GridTrackSizingAlgorithm::run()
     ASSERT(wasSetup());
     StateMachine stateMachine(*this);
 
+    if (m_renderGrid->areMasonryRows() && m_direction == ForRows)
```
Adding a helper for `RenderGrid::isMasonry(GridTrackSizingDirection)` would be nice, so that these two checks could just be `if (m_renderGrid->isMasonry(m_direction)) return;`

mattwoodrow commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1335,7 +1335,38 @@ void GridTrackSizingAlgorithm::accumulateIntrinsicSizesForTrack(GridTrack& track
+        // - Items placed at the first implicit line in the masonry axis.
+        // - Items that have a specified definite placement in the grid axis.
+        // - Items that span all grid axis tracks.
+        if (m_renderGrid->areMasonryRows() || m_renderGrid->areMasonryColumns()) {
```
Since we've always skipped sizing for a masonry axis, this check if functionally checking if the 'other' axis is masonry.

Might be nice as `m_renderGrid->isMasonry(m_direction == ForRows ? ForColumns : ForRows)`

mattwoodrow commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1335,7 +1335,38 @@ void GridTrackSizingAlgorithm::accumulateIntrinsicSizesForTrack(GridTrack& track
+        // - Items that span all grid axis tracks.
+        if (m_renderGrid->areMasonryRows() || m_renderGrid->areMasonryColumns()) {
+            bool skipTrackSizing = true;
+            auto gridAxisDirection = m_renderGrid->areMasonryRows() ? ForColumns : ForRows;
```
This is just m_direction I think.

mattwoodrow commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1335,7 +1335,38 @@ void GridTrackSizingAlgorithm::accumulateIntrinsicSizesForTrack(GridTrack& track
+        if (m_renderGrid->areMasonryRows() || m_renderGrid->areMasonryColumns()) {
+            bool skipTrackSizing = true;
+            auto gridAxisDirection = m_renderGrid->areMasonryRows() ? ForColumns : ForRows;
+            auto masonryAxisDirection = m_renderGrid->areMasonryRows() ? ForRows : ForColumns;
```
And this is just the opposite of m_direction (which I'm surprised we don't have a helper for).

mattwoodrow commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1335,7 +1335,38 @@ void GridTrackSizingAlgorithm::accumulateIntrinsicSizesForTrack(GridTrack& track
+                skipTrackSizing = false;
+
+            // Items that span all grid axis tracks.
+            if (!m_renderGrid->gridSpanForChild(*gridItem, masonryAxisDirection).startLine() && m_renderGrid->gridSpanForChild(*gridItem, gridAxisDirection).endLine() == tracks(m_direction).size())
```
This should be gridAxisDirection for both calls, right?

Could also be `m_renderGrid->gridSpanForChild(*gridItem, gridAxisDirection).integerSpan() == tracks(m_direction).size()`

mattwoodrow commented on Source/WebCore/rendering/RenderGrid.cpp:
```diff
@@ -325,8 +342,7 @@ void RenderGrid::layoutBlock(bool relayoutChildren, LayoutUnit)
         // available free space for Content Alignment.
-        if (!hasDefiniteLogicalHeight)
-            m_trackSizingAlgorithm.setFreeSpace(ForRows, logicalHeight() - trackBasedLogicalHeight);
+        m_trackSizingAlgorithm.setFreeSpace(ForRows, logicalHeight() - trackBasedLogicalHeight);
```
Rather than removing the check for `!hasDefiniteLogicalHeight`, can you just add `areMasonryRows` to it instead?

Seems less likely to break existing grid code that way.

stwrt replied:
Fixed

mattwoodrow reviewed (commented):
Looking good!

I think we need to get the tests sorted at least, since they're going to be upstreamed.

mattwoodrow commented on LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/track-sizing/masonry-track-sizing-overflow-right-side.html:
```diff
@@ -0,0 +1,57 @@
+    <title>CSS Grid Test: Verify correct behavior of track-sizing</title>
+    <link rel="author" title="Brandon Stewart" href="mailto:brandonstewart@apple.com">
+    <link rel="help" href="https://drafts.csswg.org/css-grid-3/#track-sizing">
+    <link rel="match" href="masonry-track-sizing-overflow-right-side.html">
```
Missing "-ref" suffix on both of these test file "match" headers.

mattwoodrow commented on LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/track-sizing/masonry-track-sizing-overflow-right-side-ref.html:
```diff
@@ -0,0 +1,56 @@
+    <title>CSS Grid Test: Verify correct behavior of track-sizing</title>
+    <link rel="author" title="Brandon Stewart" href="mailto:brandonstewart@apple.com">
+    <link rel="help" href="https://drafts.csswg.org/css-grid-3/#track-sizing">
+    <link rel="match" href="masonry-track-sizing-overflow-right-side-ref.html">
```
You shouldn't need a match header on the ref file. The match header is so that the test file can find the associated ref file to compare against.

mattwoodrow commented on LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/track-sizing/masonry-track-sizing-overflow-right-side.html:
```diff
@@ -0,0 +1,57 @@
+    
+<head>
+    <meta charset="utf-8">
+    <title>CSS Grid Test: Verify correct behavior of track-sizing</title>
```
Maybe "CSS Masonry Test: Items not placed at the first implicit line in the masonry axis don't contribute to track sizing" ?

mattwoodrow replied:
It would also be nice to have tests for the other two conditions in the track-sizing sections.

Have a test with items that aren't at the first line in the masonry axis, but do have an explicit position in the grid axis, and confirm that they then contribute to the track sizing again.

mattwoodrow reviewed (approved):
Looks good.

I highly recommend uploading the tests to the WPT repo, and letting the automated checks run there, before merging this, just incase those find any issues.

stwrt committed:
Implement Masonry Track Sizing
https://bugs.webkit.org/show_bug.cgi?id=248287

Reviewed by Matt Woodrow.

Track sizing will only include blocks if they meet the below criteria in the track sizing algorithm.

- Items placed at the first implicit line in the masonry axis.
- Items that have a specified definite placement in the grid axis
- Items that span all grid axis tracks.

The masonry calling code needed to be heavily refactored to a later step in layout pipeline.
This ended breaking quite a few test cases, which will be addressed in future patches.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/track-sizing/masonry-track-sizing-overflow-left-side-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/track-sizing/masonry-track-sizing-overflow-left-side.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/track-sizing/masonry-track-sizing-overflow-right-side-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/track-sizing/masonry-track-sizing-overflow-right-side.html: Added.
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/track-sizing/masonry-track-sizing-overflow-left-side-expected.txt: Added.
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/track-sizing/masonry-track-sizing-overflow-right-side-expected.txt: Added.
* Source/WebCore/rendering/GridMasonryLayout.h:
(WebCore::GridMasonryLayout::gridGap const):
* Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
(WebCore::GridTrackSizingAlgorithm::accumulateIntrinsicSizesForTrack):
(WebCore::GridTrackSizingAlgorithm::run):
* Source/WebCore/rendering/RenderGrid.cpp:
(WebCore::RenderGrid::layoutBlock):
(WebCore::RenderGrid::placeItemsOnGrid):
(WebCore::RenderGrid::autoPlacementMinorAxisDirection const):

Canonical link: https://commits.webkit.org/258280@main
